lighthouse-core/gather/driver.js

```diff
@@ -464,6 +466,30 @@ class Driver {
     }
   }
 
+  /**
+   * @typedef {any[]} TExtendsArray
+   */
+
+  /**
+   * Evaluate an expression (optionally defined in a structured manner, see `createEvalCode`
+   * in `eval.js`).
+   * @see createEvalCode
+   * @see evaluateAsync
+   * @template {TExtendsArray} T, R
+   * @param {string | ((...args: T) => R)} expressionOrMainFn
+   * @param {{useIsolation?: boolean, args?: T, deps?: Array<Function|string>}=} options
+   * @return {Promise<R>}
+   */
+  async evaluate(expressionOrMainFn, options = {}) {
+    if (typeof expressionOrMainFn !== 'string') {
+      expressionOrMainFn = createEvalCode(expressionOrMainFn, {
+        mode: 'iife',
+        ...options,
+      });
+    }
+    return this.evaluateAsync(expressionOrMainFn, options);
+  }
```

Comment:
ok, last pitch on this :) Having it inline means no (yet another) evaluateAsync helper method to look through to see how the whole system works, no `mode`, and no duplication of the relatively complicated types here. e.g.

```js
  /**
   * Evaluate a function in the context of the current page. If `useIsolation` is true, the function
   * will be evaluated in a content script that has access to the page's DOM but whose JavaScript state
   * is completely separate.
   * Returns a promise that resolves on a value of `mainFn`'s return type.
   * @template {any[]} T, R
   * @param {(...args: T) => R} mainFn The main function to call.
   * @param {{useIsolation?: boolean, args?: T, deps?: Array<Function|string>}=} options `args` should
   *   match the args of `mainFn`, and can be any serializable value. `deps` are functions that must be
   *   defined for `mainFn` to work.
   * @return {Promise<R>}
   */
  async evaluate(mainFn, options = {}) {
    const argsSerialized = options.args ? options.args.map(a => JSON.stringify(a)).join(',') : '';
    const depsSerialized = options.deps ? options.deps.join('\n') : '';

    const expression = `(() => {
      ${depsSerialized}
      ${mainFn}
      return ${mainFn.name}(${argsSerialized});
    })()`;

    return this.evaluateAsync(expression, options);
  }
```
(docs stolen from `evaluateAsync` and `createEvalCode`)

Reply:
interestingly, since `args` is optional, a caller can leave them out even if required and be ok with the type checker. I'm not sure if there's a way around that except
- requiring `options`/`args` (so all callers would have to at least provide an empty `[]` when no args are needed)
- or do it with rest params like how we do `sendCommand`, and we know the pain that causes

Reply:
I'm resisting splitting this up because it's harder to test through runtime.evaluate sniffing than directly on the string generated.

> interestingly, since args is optional ...

I tried making a conditional type that checks for tuple length and returns `{args?: T} : {}` or some variation, and that part worked fine enough, but getting the default values for the options params to work proved difficult, so I gave up and just went with requiring an empty args array.

Reply:
> I'm resisting inlining this because it's harder to test through runtime.evaluate sniffing than directly on the string generated.

But it's the same string, just surrounded by `_evaluateInContext` pre- and post-text :) (and if we replaced `window` in `_evaluateInContext` with `globalThis`, we could even `eval()` that code too).

Reply:
> but getting the default values for the options params to work proved difficult

yes! it gets weird with a `never` appearing and really mucking up the works